Add some safeguards to FutureNCCL

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48790 Add support for async callbacks in ivalue::Future
* #49014 Drop FutureNCCL in favor of vanilla CUDAFuture
* #48789 Make CUDAFuture remember and restore current device in callback
* #48840 Remove DataPtr extractor from CUDAFuture
* #48788 Cache the DataPtrs in CUDAFuture
* #48506 Split out reusable CUDAFuture from FutureNCCL
* #48505 Merge common parts of FutureNCCL into at::ivalue::Future
* #48504 Split FutureNCCL's CUDA-specific parts from generic future logic
* #48502 Support wider range of types in FutureNCCL
* #48501 Don't store device indices separately on FutureNCCL
* #48500 Add multi-GPU support to FutureNCCL
* #48563 Fix FutureNCCL not recording dataptrs with caching alloc in wait()
* #48503 Fix FutureNCCL's completed() disagreeing with wait()
* #48499 Record CUDA events for "follow-up" FutureNCCL inside markCompleted
* #48498 Use fresh stream from pool for each FutureNCCL callback
* #48497 Make FutureNCCL record events in current stream
* #48496 Have FutureNCCL record streams w/ allocator in addCallback
* **#48562 Add some safeguards to FutureNCCL**
* #48495 Remove NCCL dependency from PythonFutureWrapper
* #48561 Avoid using FutureNCCL before it's ready

This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

In this commit I'm adding a few asserts to the constructors of FutureNCCL to make sure that what's passed in is what we expect (fun fact: until two commits ago that wasn't the case, as we were passed some empty events).

I'm also making the second constructor private, as it's only supposed to be used by the then() method.

Differential Revision: [D25210333](https://our.internmc.facebook.com/intern/diff/D25210333/)

lw committed:
Add some safeguards to FutureNCCL

This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

In this commit I'm adding a few asserts to the constructors of FutureNCCL to make sure that what's passed in is what we expect (fun fact: until two commits ago that wasn't the case, as we were passed some empty events).

I'm also making the second constructor private, as it's only supposed to be used by the then() method.

Differential Revision: [D25210333](https://our.internmc.facebook.com/intern/diff/D25210333/)

[ghstack-poisoned]

lw committed:
Update on "Add some safeguards to FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

In this commit I'm adding a few asserts to the constructors of FutureNCCL to make sure that what's passed in is what we expect (fun fact: until two commits ago that wasn't the case, as we were passed some empty events).

I'm also making the second constructor private, as it's only supposed to be used by the then() method.

Differential Revision: [D25210333](https://our.internmc.facebook.com/intern/diff/D25210333/)

[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -231,8 +231,16 @@ class ProcessGroupNCCL : public ProcessGroup {
+        TORCH_INTERNAL_ASSERT(event.isCreated());
+        TORCH_INTERNAL_ASSERT(event.device_index() == deviceIndex_);
+      }
+      for (const at::DataPtr& data_ptr : extractDataPtrs(value_)) {
```
nit: (this should not block this PR, please ignore for now and fix if necessary later) can `extractDataPtrs` be expensive (e.g., using pickling)? If so, do we need to cache the extracted data ptrs?

lw replied:
My rationale was that the cost of "extracting" the data ptrs should be of the same order as "iterating" over them (after extraction). What I mean is that extracting data ptrs should have a linear complexity in the data size, and thus when combined with the linear-time iteration over the data ptrs, it doesn't effectively contribute to the overall (asymptotical) complexity.

I do realize though that this only holds when the data contains only (or mostly) tensors. If the value is a big user class with only one tensor, it won't be the case. And indeed we've ended up extracting data ptrs multiple times, so it makes sense to me to cache them. I'll add this as a separate diff.

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -231,8 +231,16 @@ class ProcessGroupNCCL : public ProcessGroup {
           "FutureNCCL only supports single-process single-device mode.");
+      for (const at::cuda::CUDAEvent& event : *cudaEvents_) {
+        TORCH_INTERNAL_ASSERT(event.isCreated());
+        TORCH_INTERNAL_ASSERT(event.device_index() == deviceIndex_);
```
I might miss sth. Why all events are on the same device? For multi-input collective calls, aren't we going to get one event per device?

lw replied:
At this stage there is in fact a single event, which is on deviceIndex_ (see the check just above). I had no real reason to use a `for` loop here rather than just check the first (and only) element of `cudaEvents_`. If you want I can change this.

mrshenli reviewed (approved):
LGTM!

lw committed:
Update on "Add some safeguards to FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

In this commit I'm adding a few asserts to the constructors of FutureNCCL to make sure that what's passed in is what we expect (fun fact: until two commits ago that wasn't the case, as we were passed some empty events).

I'm also making the second constructor private, as it's only supposed to be used by the then() method.

Differential Revision: [D25210333](https://our.internmc.facebook.com/intern/diff/D25210333/)

[ghstack-poisoned]

lw committed:
Update on "Add some safeguards to FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

In this commit I'm adding a few asserts to the constructors of FutureNCCL to make sure that what's passed in is what we expect (fun fact: until two commits ago that wasn't the case, as we were passed some empty events).

I'm also making the second constructor private, as it's only supposed to be used by the then() method.

Differential Revision: [D25210333](https://our.internmc.facebook.com/intern/diff/D25210333/)

[ghstack-poisoned]

lw committed:
Update on "Add some safeguards to FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

In this commit I'm adding a few asserts to the constructors of FutureNCCL to make sure that what's passed in is what we expect (fun fact: until two commits ago that wasn't the case, as we were passed some empty events).

I'm also making the second constructor private, as it's only supposed to be used by the then() method.

Differential Revision: [D25210333](https://our.internmc.facebook.com/intern/diff/D25210333/)

[ghstack-poisoned]

lw committed:
Update on "Add some safeguards to FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

In this commit I'm adding a few asserts to the constructors of FutureNCCL to make sure that what's passed in is what we expect (fun fact: until two commits ago that wasn't the case, as we were passed some empty events).

I'm also making the second constructor private, as it's only supposed to be used by the then() method.

Differential Revision: [D25210333](https://our.internmc.facebook.com/intern/diff/D25210333/)

[ghstack-poisoned]
